### PR TITLE
RPC: Return transaction fee from testmempoolaccept

### PR DESCRIPTION
**Pull request by rajarshimaitra**
This solves the [issue #19057. ](https://github.com/bitcoin/bitcoin/issues/19057)

A previous PR made the initial work but is removed due to commit history corruption from my end. Sadly the github discussion is also lost with the previous PR, though it was not extensive.

This PR makes the changes incorporating marcofalk's last comment of implementing `CAmount*` inside `ATMPArgs` and is ready for review.

***Note***: `testmempoolaccept` only returns `fee` if transaction is accepted in mempool.

### CODE REVIEW

**Comment by jonatack** [comment]
ISTM there is no need to close #19063 and open another PR. Better to have rebased on the original PR.

**Comment by rajarshimaitra** [comment]
I am not sure how to do that. The branch for that pr in my repo is gone. That PR somehow got a ton of commit the last time i tried to push. I tried to fix it, but then went to much more messy, so i opened a new one.

**Comment by rajarshimaitra** [comment]
I am not sure how to do that. The branch for that pr in my repo is gone. That PR somehow got a ton of commit the last time i tried to push. I tried to fix it, but then became much more messy, so i opened a new one.

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -916,10 +917,11 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
 
     TxValidationState state;
     bool test_accept_res;
+    CAmount fee = 0;
```
Comment (MarcoFalke) [inline src/rpc/rawtransaction.cpp:920]:
```suggestion
    CAmount fee_out;
```

No need to initialize. Also could clarify that this is an out parameter?

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -933,6 +935,10 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
             result_0.pushKV("reject-reason", state.GetRejectReason());
         }
     }
+    // Push tx fee, only if test_accept_res == true
+    else {
```
Comment (MarcoFalke) [inline src/rpc/rawtransaction.cpp:939]: double negatives are hard to follow (probably explains why you had to add a dev comment). Easier would be to write

```cpp
if (test_accept_res) {
 push fee
} else {
 puls reject reason
}
```

for self-documenting code

File: src/validation.cpp [hunk 3]
```diff
@@ -472,6 +472,7 @@ class MemPoolAccept
          */
         std::vector<COutPoint>& m_coins_to_uncache;
         const bool m_test_accept;
+        CAmount* fee;
```
Comment (MarcoFalke) [inline src/validation.cpp:475]:
```suggestion
        CAmount* m_fee;
```

see dev notes

File: src/validation.cpp [hunk 4]
```diff
@@ -1056,10 +1063,10 @@ bool MemPoolAccept::AcceptSingleTransaction(const CTransactionRef& ptx, ATMPArgs
 /** (try to) add transaction to memory pool with a specified acceptance time **/
 static bool AcceptToMemoryPoolWithTime(const CChainParams& chainparams, CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx,
                         int64_t nAcceptTime, std::list<CTransactionRef>* plTxnReplaced,
-                        bool bypass_limits, const CAmount nAbsurdFee, bool test_accept) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+                        bool bypass_limits, const CAmount nAbsurdFee, bool test_accept, CAmount* fee_out=nullptr) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
Comment (MarcoFalke) [inline src/validation.cpp:1066]: please move the fee arg before the test arg, so that non-testing callers can use this as well (without having to specify that they don't test)

File: test/functional/mempool_accept.py [hunk 5]
```diff
@@ -104,7 +104,7 @@ def run_test(self):
         ))['hex']
         tx.deserialize(BytesIO(hex_str_to_bytes(raw_tx_final)))
         self.check_mempool_result(
-            result_expected=[{'txid': tx.rehash(), 'allowed': True}],
+            result_expected=[{'txid': tx.rehash(), 'allowed': True, 'fee': (int(coin['amount']) - (0.025)) * COIN}],
```
Comment (MarcoFalke) [inline test/functional/mempool_accept.py:107]: please no floating point in tests. Decimal is good enough

File: test/functional/mempool_accept.py [hunk 6]
```diff
@@ -187,7 +187,7 @@ def run_test(self):
         tx.deserialize(BytesIO(hex_str_to_bytes(raw_tx_reference)))
         # Reference tx should be valid on itself
         self.check_mempool_result(
-            result_expected=[{'txid': tx.rehash(), 'allowed': True}],
+            result_expected=[{'txid': tx.rehash(), 'allowed': True, 'fee': (0.1 - 0.05) * COIN}],
```
Comment (MarcoFalke) [inline test/functional/mempool_accept.py:190]: Same

**Review by maflcko — COMMENTED** [review]
Concept ACK

File: src/rpc/rawtransaction.cpp (hunk 1 above)
Comment (rajarshimaitra) [reply src/rpc/rawtransaction.cpp:920]: Done.

File: src/rpc/rawtransaction.cpp (hunk 2 above)
Comment (rajarshimaitra) [reply src/rpc/rawtransaction.cpp:939]: makes sense, done.

File: src/validation.cpp (hunk 3 above)
Comment (rajarshimaitra) [reply src/validation.cpp:475]: My mistake. Fixed.

File: src/validation.cpp (hunk 4 above)
Comment (rajarshimaitra) [reply src/validation.cpp:1066]: Done.

File: test/functional/mempool_accept.py (hunk 5 above)
Comment (rajarshimaitra) [reply test/functional/mempool_accept.py:107]: I am a little confused here, i used `int` because `Decimal` wasn't asserting with `int` type return value in fee. `0.025` is the output value in btc which is already used few lines above while constructing the tx. What exactly you are suggesting to change here?

Comment (MarcoFalke) [reply test/functional/mempool_accept.py:107]: Nevemind. I wrote this test and used floating point in the first place. :sweat_smile: :cold_sweat:

File: src/validation.cpp (hunk 4 above)
Comment (rajarshimaitra) [reply src/validation.cpp:1066]: Also i was wondering cant a non-testing call use something `(..., false, &fee)` to use the `fee` while passing `false` in `test_accept`?

Comment (MarcoFalke) [reply src/validation.cpp:1066]: Ok, fair enough. No need to change if it is too involved.

**Comment by rajarshimaitra** [comment]
Thanks for the review. Updated and rebased.

**Review by promag — COMMENTED** [review]
Concept ACK, code change looks good.

Could add release note and also add/update a test without `fee` in the result.

**Comment by rajarshimaitra** [comment]
Thanks @promag for the review and comment. 
I am not very familiar with release notes update protocols, is this something that I should change as part of this PR? 
Also, can you elaborate a little on the testing part that you mentioned?

**Comment by luke-jr** [comment]
https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#release-notes

https://github.com/bitcoin/bitcoin/blob/master/test/functional/README.md

File: src/rpc/rawtransaction.cpp [hunk 7]
```diff
@@ -916,13 +917,18 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
...
-            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true);
+            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true, &fee);
     }
     result_0.pushKV("allowed", test_accept_res);
-    if (!test_accept_res) {
+    if (test_accept_res) {
+        // Push tx fee, only if test_accept_res == true
+        result_0.pushKV("fee", fee);
+    }
+    else {
```
Comment (luke-jr) [inline src/rpc/rawtransaction.cpp:931]: Should be on one line

File: src/validation.cpp [hunk 8]
```diff
@@ -683,6 +684,12 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
         return error("%s: Consensus::CheckTxInputs: %s, %s", __func__, tx.GetHash().ToString(), state.ToString());
     }
 
+    // If fee_out is passed, return the fee to the caller
+    if (args.m_fee != nullptr){
```
Comment (luke-jr) [inline src/validation.cpp:688]:
```suggestion
    if (args.m_fee) {
```

**Review by luke-jr — APPROVED** [review]
utACK, with some nits

File: src/rpc/rawtransaction.cpp (hunk 7 above)
Comment (rajarshimaitra) [reply src/rpc/rawtransaction.cpp:931]: Done.

File: src/validation.cpp (hunk 8 above)
Comment (rajarshimaitra) [reply src/validation.cpp:688]: Done.

**Comment by rajarshimaitra** [comment]
Thanks @luke-jr  for the review. Updated and rebased. 

@promag I have added a release note. Let me know if any modification needed there. 

Regarding the extra functional test, `testmempoolaccept` will always return the fee as long as the transaction is accepted in mempool. Fee will only be absent from the result if transaction acceptance check fails, which is already covered in the existing tests. So if I understand correctly, these two cases (with fee, without fee) are being checked already. Let me know if you think any more cases should need testing.

File: doc/release-notes-19093.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+RPC changes
+------------
+
+### 'testmempoolaccept` imporvement
```
Comment (luke-jr) [inline doc/release-notes-19093.md:4]: Typo

File: src/validation.h [hunk 10]
```diff
@@ -191,10 +191,11 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
 void PruneBlockFilesManual(int nManualPruneHeight);
 
 /** (try to) add transaction to memory pool
- * plTxnReplaced will be appended to with all transactions replaced from mempool **/
+ * plTxnReplaced will be appended to with all transactions replaced from mempool
+ * optionally takes an arguement to return tx fee to the caller **/
```
Comment (luke-jr) [inline src/validation.h:195]: Typo

File: doc/release-notes-19093.md (hunk 9 above)
Comment (rajarshimaitra) [reply doc/release-notes-19093.md:4]: Fixed.

File: src/validation.h (hunk 10 above)
Comment (rajarshimaitra) [reply src/validation.h:195]: Fixed

File: src/rpc/rawtransaction.cpp [hunk 11]
```diff
@@ -916,13 +917,17 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
...
+    CAmount fee; // To return transaction fee from AcceptToMemoryPool
     {
         LOCK(cs_main);
         test_accept_res = AcceptToMemoryPool(mempool, state, std::move(tx),
-            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true);
+            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true, &fee);
     }
     result_0.pushKV("allowed", test_accept_res);
-    if (!test_accept_res) {
+    if (test_accept_res) {
```
Comment (glowang) [inline src/rpc/rawtransaction.cpp:927]: Could you elaborate on the consideration behind only passing back fee rate if the transaction is accepted? Late to the party, so I might have missed it if this was asked earlier by a different reviewer.

I think we should still pass back fee even if rejected, because In validation.h, we can see that one potential reason of rejection is`TX_MEMPOOL_POLICY`, which includes reasons such as mempool min fee not met, min relay fee not met, etc (in `CheckFeeRate`). 

This RPC is also potentailly improved to process more than one transaction in the future, which means that more than one trasactions could fail. At that point, it would be helpful for users to see directly the transactions that are not accepted & the exact fees they chose for the corresponding transactions.

Comment (rajarshimaitra) [reply src/rpc/rawtransaction.cpp:927]: The rationale behind not returning fees in case of rejection was mempool policy is not the only reason why a transaction might be rejected. In case of ill-formed transactions, returning a fee might cause a problem because it might try to return something that doesn't make sense. That might add up confusion about the result than help.

On the other hand, `CheckFeeRate` already returns both the `package fee` and `min_fee*` as an error message. So that information is already available. 

So it felt redundant to try to return fee in case of failure in the RPC call and handle all the edge cases. If it needs to be done I think it's better to do it in a separate PR and it will need some added testings too to ensure what is returned actually make sense.

Comment (glowang) [reply src/rpc/rawtransaction.cpp:927]: >  If I am not mistaken then it will return 0 fee for ill-formed transactions.
@rajarshimaitra would you mind pointing me to the code to look more into it? I did some searching myself but wasn't able to find this.

Comment (rajarshimaitra) [reply src/rpc/rawtransaction.cpp:927]: Check here.

https://github.com/bitcoin/bitcoin/blob/205b87d2f6bd01285de50ba742e32e4ab1298b13/src/validation.cpp#L688-L691

if `CheckTxInputs` fails it will return `nFess` will be 0. Look inside `CheckTxInputs` to see what makes it fail.

**Review by glozow — COMMENTED** [review]
Concept ACK!

I'm wondering why you decided to return fee instead of feerate? I think we typically use feerate to decide whether a fee is reasonable, although a user might prefer to just see a fee amount. It's a quick conversion between the two and I don't have a strong opinion, but was wondering what your rationale was?

**Review by glozow — COMMENTED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/19093/commits/3218bba02ce6c0ae17111e340872f02a000f1dc3

I'm building on top of this in #19339, just fyi @rajarshimaitra

**Comment by rajarshimaitra** [comment]
>I'm wondering why you decided to return fee instead of feerate? I think we typically use feerate to decide whether a fee is reasonable, although a user might prefer to just see a fee amount. It's a quick conversion between the two and I don't have a strong opinion, but was wondering what your rationale was?

That's a good proposal. I didn't thought about it, but it makes sense to report the fee rate instead of the total fee.

I think its a better thing to do. Will change if others agree.

>I'm building on top of this in #19339, just fyi @rajarshimaitra

It looks ready from my side. Maybe needs few more ACKs.

**Comment by MarcoFalke** [comment]
I am starting to think that it could even return the same info that `getmempoolinfo` returns (except the DEPRECATED fields and the mempool entry time). (Maybe hidden behind a `verbosity` switch, if someone only wants a boolean acceptance result)

File: src/validation.cpp [hunk 12]
```diff
@@ -690,6 +691,12 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
         return error("%s: Consensus::CheckTxInputs: %s, %s", __func__, tx.GetHash().ToString(), state.ToString());
     }
 
+    // If fee_out is passed, return the fee to the caller
+    if (args.m_fee) {
+        *args.m_fee = nFees;
+    }
+
```
Comment (jnewbery) [inline src/validation.cpp:698]: no need for two empty lines here.

File: src/rpc/rawtransaction.cpp [hunk 13]
```diff
@@ -916,13 +917,17 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
 
     TxValidationState state;
     bool test_accept_res;
+    CAmount fee; // To return transaction fee from AcceptToMemoryPool
```
Comment (jnewbery) [inline src/rpc/rawtransaction.cpp:920]: nit: no need for this inline comment. It's clear from the code what this variable is doing.

File: src/rpc/rawtransaction.cpp [hunk 14]
```diff
@@ -916,13 +917,17 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
...
     {
         LOCK(cs_main);
         test_accept_res = AcceptToMemoryPool(mempool, state, std::move(tx),
-            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true);
+            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true, &fee);
     }
     result_0.pushKV("allowed", test_accept_res);
-    if (!test_accept_res) {
+    if (test_accept_res) {
+        // Push tx fee, only if test_accept_res == true
```
Comment (jnewbery) [inline src/rpc/rawtransaction.cpp:928]: No need for a comment to say what the code is doing. If you want to comment here, it should be to explain _why_ the code is doing this, eg "if mempool acceptance failed, the returned fee may not be accurate, so only return it to the caller if the tx would be accepted." or similar.

File: doc/release-notes-19093.md [hunk 15]
```diff
@@ -0,0 +1,7 @@
+RPC changes
+------------
+
+### 'testmempoolaccept` improvement
+
+`testmempoolaccept` now additionally returns the transaction fee in the result.
+Fee is only returned if transaction is accepted in the mempool.
```
Comment (jnewbery) [inline doc/release-notes-19093.md:7]: "Fee is only returned if transaction would be accepted to the mempool." is more accurate (testmempoolaccept doesn't actually accept transactions to the mempool).

**Review by jnewbery — COMMENTED** [review]
Code looks good. A few style comments inline.

Additionally, a couple of comments on the commit log:

- the summary line should be in the imperative mood (i.e. "Return transaction fee from testmempoolaccept" instead of "testmempoolaccept returns transaction fee")
- change "if the transaction is accepted in mempool." to "if the transaction would be accepted to the mempool."

File: src/rpc/rawtransaction.cpp (hunk 13 above)
Comment (rajarshimaitra) [reply src/rpc/rawtransaction.cpp:920]: Removed this comment.

File: src/rpc/rawtransaction.cpp (hunk 14 above)
Comment (rajarshimaitra) [reply src/rpc/rawtransaction.cpp:928]: Updated this comment to better express the intent.

File: doc/release-notes-19093.md (hunk 15 above)
Comment (rajarshimaitra) [reply doc/release-notes-19093.md:7]: updated.

**Comment by rajarshimaitra** [comment]
Thanks @jnewbery for the review. Updated as per your suggestions and rebased.

**Review by glozow — COMMENTED** [review]
reACK https://github.com/bitcoin/bitcoin/pull/19093/commits/c9781e92a211d2615a8889dede1b02913c18ecd1

**Comment by jnewbery** [comment]
Code review ACK c9781e92a211d2615a8889dede1b02913c18ecd1

s390 failure in travis is unrelated to this PR. I've restarted the job.

File: src/rpc/rawtransaction.cpp [hunk 16]
```diff
@@ -878,6 +878,7 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
                         {
                             {RPCResult::Type::STR_HEX, "txid", "The transaction hash in hex"},
                             {RPCResult::Type::BOOL, "allowed", "If the mempool allows this tx to be inserted"},
+                            {RPCResult::Type::NUM, "fee", "Fee provided in this transaction (only present when 'allowed' is true)"},
```
Comment (MarcoFalke) [inline src/rpc/rawtransaction.cpp:881]: * This will put the fee in raw satoshis? I don't think we do this anywhere else.
* Have you seen https://github.com/bitcoin/bitcoin/pull/19093#issuecomment-663963257? Specifically, it would be better for the caller to have the feerate instead of the absolute fee. Otherwise, the caller is forced to implement a witness-aware tx-deserializer to determine the virtual size. `getmempoolentry` returns vsize, weight, as well as a `fees` object.

**Review by maflcko — COMMENTED** [review]
Concept ACK

File: src/rpc/rawtransaction.cpp (hunk 16 above)
Comment (rajarshimaitra) [reply src/rpc/rawtransaction.cpp:881]: Hmm. Now that I am thinking it makes sense to return `feerate`. Do you want the `fees` to be changed into `feerate`? 

> I am starting to think that it could even return the same info that `getmempoolinfo` returns

Little confused on this one as `getmepoolinfo` returns the following
```
Result:
{
  "loaded": true|false         (boolean) True if the mempool is fully loaded
  "size": xxxxx,               (numeric) Current tx count
  "bytes": xxxxx,              (numeric) Sum of all virtual transaction sizes as defined in BIP 141. Differs from actual serialized size because witness data is discounted
  "usage": xxxxx,              (numeric) Total memory usage for the mempool
  "maxmempool": xxxxx,         (numeric) Maximum memory usage for the mempool
  "mempoolminfee": xxxxx       (numeric) Minimum fee rate in BTC/kB for tx to be accepted. Is the maximum of minrelaytxfee and minimum mempool fee
  "minrelaytxfee": xxxxx       (numeric) Current minimum relay fee for transactions
}
```

Which of these do you think should be included here?

Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp:881]: Sorry, typo. Should have been getmempool**entry** in the other comment as well

Comment (jnewbery) [reply src/rpc/rawtransaction.cpp:881]: This seems like scope creep. I think returning the fee is fine to start with, and unblocks #19339. Could we defer adding more details to a later PR?

Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp:881]: * Changing the type to `STR_AMOUNT` like in all other places should be done for consistency and to protect against mistakes. I consider this a bug that should be fixed before merge.
* Removing the field `fee` and replacing it with something else later is a breaking rpc change. For the sanity of our users we should keep breaking rpc changes at a minimum.

Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp:881]: Does #19339 need the rpc changes? If not, then this can trivially be unblocked by cherry-picking this pull and undoing the RPC changes. I think an internal (non-rpc) change can go in earlier, as it doesn't have the breaking rpc change concerns.

Comment (jnewbery) [reply src/rpc/rawtransaction.cpp:881]: Good point about cherry picking the non-rpc changes.

Separately, I think `fee` is a useful value to return to the user, since it's exact (unlike feerate which might be rounded).

Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp:881]: Agree, since `getmempoolentry` also returns the absolute fee (and the vsize)

Comment (gzhao408) [reply src/rpc/rawtransaction.cpp:881]: Yeah good point, I only need the fee from ATMP for #19339 so I'll just use that. @rajarshimaitra is listed as coauthor on there, and can continue iterating on the RPC stuff here.

File: src/validation.cpp [hunk 17]
```diff
@@ -475,6 +475,7 @@ class MemPoolAccept
          */
         std::vector<COutPoint>& m_coins_to_uncache;
         const bool m_test_accept;
+        CAmount* m_fee;
```
Comment (fjahr) [inline src/validation.cpp:478]: nit: to match your naming elsewhere you could have called this `m_fee_out`

**Review by fjahr — COMMENTED** [review]
Code review ACK c9781e92a211d2615a8889dede1b02913c18ecd1

Ignore my nit unless you have to retouch.

**Comment by jarrad** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
I still consider my comment a bug that should be fixed before merge. Also my proposed change would be a breaking RPC change, and I wouldn't want to expose users to that. So better get it right the first time and avoid another "lets do it later" like e.g. #19543.

**Comment by MarcoFalke** [comment]
(Ok, just adding new fields to an object is not a breaking change. Only if the new fields were wrapped into another object, but that probably doesn't make sense here)

**Comment by jnewbery** [comment]
I've chatted to Marco offline and agree that this PR should be changed to add `vsize` and `fees.base`, to be consistent with `getmempoolentry`. If we want to add other fields later (eg `fees.modified`), that's not a breaking change. We shouldn't add `fee`, since that's deprecated in `getmempoolentry`.

**Comment by gzhao408** [comment]
I agree with @MarcoFalke. 
I also had a few comments on my PR about these changes and it felt unfair to force @rajarshimaitra to follow them just because I wanted to use this code... so [here's a branch](https://github.com/gzhao408/bitcoin/commit/67b1626ca29fe895d9c552714a18ab911f4e20b1) that applies the suggestions from my PR. I incorporated Marco's comment as well, so it now returns a `fees` object with a `STR_AMOUNT` "base" field for what we get from `PreChecks` here.

Illustrated more clearly (just the relevant bits):
```
{
    RPCResult::Type::BOOL 'allowed'
    RPCResult::Type::NUM 'vsize'
    RPCResult::Type::OBJ 'fees' 
    {
        RPCResult::Type::STR_AMOUNT 'base'
    }
}
```
Summary of the small edits if you're interested: update release-notes.md instead of separate doc, doxygen comment for the optional `fee_out`, and Fabian's naming comment above.

@rajarshimaitra please feel free to use this branch :)

**Comment by rajarshimaitra** [comment]
Hi, Sorry I got busy with other works and couldn't check the updates here lately. Honestly, I am a little overwhelmed with all the discussions regarding this, couldn't really follow up on what's happening where.
So if someone can explain in simple terms what I need to do, if anything, to make this PR mergeable would be happy to make the changes and close this.

**Comment by jnewbery** [comment]
Hi @rajarshimaitra . Thanks for your work on this. It sounds like @gzhao408 has a bit more time than you to make the final changes here, so I'd suggest that she open a PR that takes your branch and makes those last minor changes to get this merged. What do you think?

**Comment by gzhao408** [comment]
> It sounds like @gzhao408 has a bit more time than you to make the final changes here, so I'd suggest that she open a PR that takes your branch and makes those last minor changes to get this merged.

Yep I can do that! @rajarshimaitra you'll be co-author of the commit (the branch adding the suggested changes is a cherry-pick from yours). Does this sound alright to you?

**Comment by rajarshimaitra** [comment]
@jnewbery and @gzhao408 yes that sounds good to me. As I have lost touch over this it would be better if she takes over and do the last changes. ACK from me.

**Comment by jnewbery** [comment]
New version is at #19940 . Thanks for your hard work on this @rajarshimaitra !

@jonatack @MarcoFalke @fjahr @promag @luke-jr - mind re-reviewing the new version at #19940? It's this branch with some minor updates.
